tools/topology/sof/pipe-src-capture.m4

```diff
@@ -2,62 +2,110 @@
 #
 # Pipeline Endpoints for connection are :-
 #
-#  host PCM_P --> SRC --> sink DAI0
+#  host PCM_C <-- SRC <-- sink DAI0
 
 # Include topology builder
 include(`utils.m4')
 include(`src.m4')
 include(`buffer.m4')
 include(`pcm.m4')
 include(`dai.m4')
+include(`pga.m4')
 include(`pipeline.m4')
+include(`mixercontrol.m4')
+
+#
+# Controls
+#
+
+# Volume Mixer control with max value of 32
+C_CONTROLMIXER(Master Capture Volume, PIPELINE_ID,
+	CONTROLMIXER_OPS(volsw,
+		256 binds the mixer control to volume get/put handlers,
+		256, 256),
+	CONTROLMIXER_MAX(, 80),
+	false,
+	CONTROLMIXER_TLV(TLV 80 steps from -50dB to +30dB for 1dB, vtlv_m50s1),
+	Channel register and shift for Front Left/Right,
+	LIST(`	', KCONTROL_CHANNEL(FL, 1, 0), KCONTROL_CHANNEL(FR, 1, 1)))
+
+
+# Volume Configuration
+define(CAPTURE_PGA_TOKENS, concat(`capture_pga_tokens_', PIPELINE_ID))
+define(CAPTURE_PGA_CONF, concat(`capture_pga_conf_', PIPELINE_ID))
+
+W_VENDORTUPLES(CAPTURE_PGA_TOKENS, sof_volume_tokens,
+LIST(`		', `SOF_TKN_VOLUME_RAMP_STEP_TYPE	"0"'
+     `		', `SOF_TKN_VOLUME_RAMP_STEP_MS		"250"'))
+
+W_DATA(CAPTURE_PGA_CONF, CAPTURE_PGA_TOKENS)
 
 #
 # Components and Buffers
 #
 
 # Host "Passthrough Capture" PCM
-# with 4 sink and 0 source periods
-W_PCM_CAPTURE(PCM_ID, Passthrough Capture, 4, 0)
+# with 0 sink and 3 source periods
+W_PCM_CAPTURE(PCM_ID, SRC Capture, 0, 3)
 
 #
 # SRC Configuration
 #
 
-W_VENDORTUPLES(media_src_tokens, sof_src_tokens, LIST(`		', `SOF_TKN_SRC_RATE_IN	"PIPELINE_RATE"'))
+# Create unique names for data to allow several different input rate instances
+define(CAPTURE_SRC_TOKENS, concat(`capture_src_tokens_', PIPELINE_ID))
+define(CAPTURE_SRC_CONF, concat(`capture_src_conf_', PIPELINE_ID))
```

Comment:
There's no undefine() for these macros. They are not needed outside this macro so better to do it.

Reply:
Done now (undefines add plus minor topology cleanup). I also created separate pipeline building block macros pipe-src-capture.m4 and pipe-src-volume-capture.m4. The latter is more generic with format convert capability. The other ASRC PR has similar pipeline macros.